main.go

```diff
@@ -39,65 +40,78 @@ func main() {
 	go runEthListener(ethEvents, &currentBlock)
 	go terminateExpiringOrgs(stateEvents, &currentBlock)
 
+	channels := map[string]OrgChannel{}
+	channelCloseEvents := make(chan string)
+
 	for {
+		select {
 		// stream events from contract
-		e := <-ethEvents
-
-		if err := db.UpsertEvent(e); err != nil {
-			l.Fatal("Failed to upsert event", e, err)
-		}
+		case e := <-ethEvents:
+			if err := db.UpsertEvent(e); err != nil {
+				l.Fatal("Failed to upsert event", e, err)
+			}
 
-		// if event has been removed, a reorg has happened
-		if e.Removed {
-			stateAfterRemoval(&e, &currentBlock)
-		}
+			// if event has been removed, a reorg has happened
+			if e.Removed {
+				stateAfterRemoval(&e, &currentBlock)
+			}
 
-		if queue, ok := queues.LoadOrStore(e.Org, utils.NewEventQueue(e)); ok {
-			// queue already exists, and a goroutine is running, just add event
-			q := queue.(*utils.EventQueue)
-			q.AddEvent(e)
-			// between LoadOrStore and AddEvent, queue might get deleted from map
-			if _, ok := queues.LoadOrStore(e.Org, q); !ok {
-				// re-spawn another goroutine for this org
-				go processEventsForOrg(e.Org, stateEvents)
+			ch := getOrCreateOrgChannel(e.Org, &channels)
+			ch.event <- e
+			select {
+			case <-ch.ack:
+			case <-time.After(3 * time.Second):
+				// if we didn't get the ack, we assume the processor is not running
+				go processEventsForOrg(e.Org, ch, stateEvents, channelCloseEvents)
```

Comment:
Why not start the goroutine when the org is added to the map?

Reply:
That's what we did in https://github.com/radicle-dev/radicle-cloud/commit/e25f469764430d8b9fb72f901c1a9436925e30af#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R57-R59 since we know only one `'static` goroutine would be running.

Here we don't know whether it was terminated or not.